geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/collections/OrderStatisticTreeTest.java

```diff
@@ -0,0 +1,580 @@
+          assertThatThrownBy(() -> iterator2.remove()).isInstanceOf(IllegalStateException.class);
+        }
+      } else {
+        assertEquals(iterator1.hasNext(), iterator2.hasNext());
+
+        if (iterator1.hasNext()) {
+          assertEquals(iterator1.next(), iterator2.next());
+        } else {
+          break;
+        }
+      }
+    }
+
+    assertEquals(set.size(), tree.size());
+    assertTrue(tree.isHealthy());
+    assertTrue(set.containsAll(tree));
+    assertTrue(tree.containsAll(set));
+  }
+
+  @Test
+  public void testIteratorConcurrentModification() {
+    for (int i = 0; i < 100; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> iterator1 = set.iterator();
+    Iterator<Integer> iterator2 = tree.iterator();
+
+    set.remove(10);
+    tree.remove(10);
+
+    assertEquals(iterator1.hasNext(), iterator2.hasNext());
+
+    try {
+      iterator1.next();
+      try {
+        iterator2.next();
+      } catch (ConcurrentModificationException ex) {
+        fail("iterator2 should NOT have thrown an exception.");
+      }
+    } catch (ConcurrentModificationException ex) {
+      assertThatThrownBy(() -> iterator2.next())
+          .isInstanceOf(ConcurrentModificationException.class);
+    }
+  }
+
+
+  @Test
+  public void testIteratorConcurrentRemove() {
+    for (int i = 10; i < 20; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> iterator1 = set.iterator();
+    Iterator<Integer> iterator2 = tree.iterator();
+
+    for (int i = 0; i < 4; ++i) {
+      iterator1.next();
+      iterator2.next();
+    }
+
+    // None of them contains 2, should not change the modification count.
+    set.remove(2);
+    tree.remove(2);
+
+    iterator1.remove();
+    iterator2.remove();
+
+    iterator1.next();
+    iterator2.next();
+
+    set.remove(12);
+    tree.remove(12);
+
+    // Both of them should throw.
+    assertThatThrownBy(() -> iterator1.remove())
+        .isInstanceOf(ConcurrentModificationException.class);
+    assertThatThrownBy(() -> iterator2.remove())
+        .isInstanceOf(ConcurrentModificationException.class);
+  }
+
+  @Test
+  public void testConcurrentOrIllegalStateOnRemove() {
+    for (int i = 0; i < 10; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> setIterator = set.iterator();
+    Iterator<Integer> treeIterator = tree.iterator();
+
+    set.add(100);
+    tree.add(100);
+
+    assertThatThrownBy(() -> setIterator.remove()).isInstanceOf(IllegalStateException.class);
+    assertThatThrownBy(() -> treeIterator.remove()).isInstanceOf(IllegalStateException.class);
+  }
+
+  @Test
+  public void testConcurrentIterators() {
+    for (int i = 0; i < 10; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> iterator1a = set.iterator();
+    Iterator<Integer> iterator1b = set.iterator();
+    Iterator<Integer> iterator2a = tree.iterator();
+    Iterator<Integer> iterator2b = tree.iterator();
```

Comment:
Could these be renamed to something more descriptive please.

Reply:
Renamed.